Acme v2

unixcharles opened the pull request:
Lot of changes here.

Most of the concept remain the same but its almost a complete rewrite for v2. Some API change were made for both consistency but also to stay as close as possible to the specification.

Notable design changes

* Resource objects are now easy to serialize and re-instantiate.
* Resource objects can all be retrieve from URL.
* Resource objects delegate API requests back to the client itself.
* The client does not make any assumption on API endpoints and retrieve everything from the directory URL.
* Group error, client errors vs server errors.

I also re-written documentation / readme.

The API goal for the gem is to stay close to the protocol itself while having this Ruby feeling that make it easy to work with.

will-in-wi commented on README.md:
```diff
@@ -21,135 +21,171 @@ gem 'acme-client'
-# You should probably use the staging endpoint for all your experimentation:
-# endpoint = 'https://acme-staging.api.letsencrypt.org/'
-endpoint = 'https://acme-v01.api.letsencrypt.org/'
+To order a new certificate, the client must provider a list of identifiers.
```
"provider" to "provide"?

will-in-wi commented on README.md:
```diff
@@ -21,135 +21,171 @@ gem 'acme-client'
-# Initialize the client
-require 'acme-client'
-client = Acme::Client.new(private_key: private_key, endpoint: endpoint, connection_options: { request: { open_timeout: 5, timeout: 5 } })
+The returned order will contains a list of `Authorization` that need to be completed in other to finalize the order, generally one per identifiers.
```
"contains" to "contain"?

Since `Authorization` is singular and the context is plural, would it make sense to phrase it like "…contain a list of `Authorization` objects that need…"

"identifiers" to "identifier"?

will-in-wi commented on README.md:
```diff
@@ -21,135 +21,171 @@ gem 'acme-client'
 
-# If the private key is not known to the server, we need to register it for the first time.
-registration = client.register(contact: 'mailto:contact@example.com')
+Each authorization contains multiple challenges, typically a `dns-01` and a `http-01` challenge. The applicant is only required to complete one of the two challenge.
```
"challenge" to "challenges"

will-in-wi commented on README.md:
```diff
@@ -21,135 +21,171 @@ gem 'acme-client'
+To complete the HTTP challenge, you must prepare return a file using HTTP.
 
-Before you are able to obtain certificates for your domain, you have to prove that you are in control of it.
+The path follow the following format:
```
"follow" to "follows"

will-in-wi commented on README.md:
```diff
@@ -21,135 +21,171 @@ gem 'acme-client'
-### Authorize for domain
+### Preparing for HTTP challenge
+
+To complete the HTTP challenge, you must prepare return a file using HTTP.
```
Should "prepare" be removed? E.g. "…you must return a file using HTTP."

will-in-wi commented on README.md:
```diff
@@ -21,135 +21,171 @@ gem 'acme-client'
+
+> .well-known/acme-challenge/#{token}
+
+And the file content is the key authorization. The HTTP01 object as utility methods to generate them.
```
"The HTTP01 object **has** utility methods"

will-in-wi commented on README.md:
```diff
@@ -21,135 +21,171 @@ gem 'acme-client'
-# If authorization.status returns 'valid' here you can already get a certificate
-# and _must not_ try to solve another challenge.
-authorization.status # => 'pending'
+For test purpose you can just save the challenge file and use Ruby to serve it:
```
"purpose" to "purposes"

will-in-wi commented on README.md:
```diff
@@ -21,135 +21,171 @@ gem 'acme-client'
 
-# You can retrieve the challenge token
-challenge.token # => "some_token"
+The DNS01 object as utility methods to generate them.
```
"DNS01 object **has** utility methods"

will-in-wi commented on README.md:
````diff
@@ -21,135 +21,171 @@ gem 'acme-client'
+```ruby
+> dns_challenge.record_name # => '_acme-challenge'
+> dns_challenge.record_type # => 'TXT'
+> dns_challenge.record_content # => 'HRV3PS5sRDyV-ous4HJk4z24s5JjmUTjcCaUjFt28-8'
````
You might not want the `> ` at the start of the lines, since it inhibits copying and is inconsistent with the rest of the README.

will-in-wi commented on README.md:
```diff
@@ -21,135 +21,171 @@ gem 'acme-client'
 
-# You are not required to send a Content-Type. This method will return the right Content-Type should you decide to include one.
-challenge.content_type
+Once your are ready to complete the challenge, you can request to server to perform the verification.
```
"your are" to "you are"

Perhaps "you can request the server perform the verification"

will-in-wi commented on README.md:
```diff
@@ -21,135 +21,171 @@ gem 'acme-client'
 
-# Wait a bit for the server to make the request, or just blink. It should be fast.
-sleep(1)
+CSR can be slightly tricky to generate using OpenSSL from Ruby standard library. The `acme-client` provide a utility class `CertificateRequest` to help with that.
```
"A CSR"

Perhaps "`acme-client` provides a utility class" (no "The")

will-in-wi commented on README.md:
```diff
@@ -21,135 +21,171 @@ gem 'acme-client'
-# the challenge is not relevant and in fact may never change from pending.
-challenge.authorization.verify_status # => 'valid'
-challenge.error # => nil
+The certificate generation is happening asynchronously. You may need to poll.
```
Perhaps "Certificate generation happens asynchronously"?

will-in-wi commented on README.md:
```diff
@@ -21,135 +21,171 @@ gem 'acme-client'
+order.finalize(csr: csr)
+until order.status != 'processing'
+  sleep(1)
+end
```
Style preference, but I think you could do this as a one-liner:

```ruby
sleep(1) while order.status == 'processing'
```

Also, `while` might be clearer than `until`, given that there is essentially a double negative between `until` and `!=`.

will-in-wi commented on README.md:
```diff
@@ -21,135 +21,171 @@ gem 'acme-client'
+### Certificate revokation
 
-Now that your account is authorized for the domain, you should be able to obtain a certificate for it.
+To revoke a certificate you can call `#revoke` with certificate.
```
"can call `#revoke` with **the** certificate"

will-in-wi commented on README.md:
```diff
@@ -21,135 +21,171 @@ gem 'acme-client'
-# Not implemented
+### Certificate renewal
+
+The is no renewal precess, just create a new order.
```
"precess" to "process"

will-in-wi commented on lib/acme/client.rb:
```diff
@@ -22,15 +21,14 @@ class Acme::Client; end
-  def initialize(jwk: nil, private_key: nil, endpoint: DEFAULT_ENDPOINT, directory_uri: nil, connection_options: {})
+  DEFAULT_DIRECTORY = 'http://127.0.0.1:4000/directory'.freeze
+  repo_url = 'https://github.com/unixcharles/acme-client'
+  USER_AGENT = "Acme::Client v#{Acme::Client::VERSION} (#{repo_url})".freeze
```
It looks like `repo_url` is only used in `USER_AGENT`. If this is the case, would it make more sense to just interpolate it?

unixcharles replied:
Its just to keep the line short

will-in-wi commented on lib/acme/client.rb:
```diff
@@ -41,93 +39,226 @@ def initialize(jwk: nil, private_key: nil, endpoint: DEFAULT_ENDPOINT, directory
-    response = connection.post(@operation_endpoints.fetch('new-reg'), payload)
-    ::Acme::Client::Resources::Registration.new(self, response)
+    if terms_of_service_agreed
+      payload['termsOfServiceAgreed'] = terms_of_service_agreed
```
Why are the the keys strings here and symbols in `#account_update`?

will-in-wi commented on lib/acme/client.rb:
```diff
@@ -41,93 +39,226 @@ def initialize(jwk: nil, private_key: nil, endpoint: DEFAULT_ENDPOINT, directory
+    if @kid.nil?
+      response = post(endpoint_for('newAccount'), payload: { onlyReturnExisting: true }, mode: :jwk)
+      @kid = response.headers.fetch(:location)
+    end
```
What about

```ruby
@kid ||= begin
  response = post(endpoint_for('newAccount'), payload: { onlyReturnExisting: true }, mode: :jwk)
  response.headers.fetch(:location)
end
```

or even

```ruby
@kid ||= begin
  post(endpoint_for('newAccount'), payload: { onlyReturnExisting: true }, mode: :jwk)
    .headers
    .fetch(:location)
end
```

will-in-wi commented on lib/acme/client.rb:
```diff
@@ -41,93 +39,226 @@ def initialize(jwk: nil, private_key: nil, endpoint: DEFAULT_ENDPOINT, directory
+
+  def finalize(url:, csr:)
+    unless csr.respond_to?(:to_der)
+      raise ArgumentError, 'csr must respond to `to_der`'
```
Maybe change the error message to `#to_der`?

will-in-wi commented on lib/acme/client.rb:
```diff
@@ -41,93 +39,226 @@ def initialize(jwk: nil, private_key: nil, endpoint: DEFAULT_ENDPOINT, directory
+    base64_der_csr = Acme::Client::Util.urlsafe_base64(csr.to_der)
+
+    payload = { csr: base64_der_csr }
+    response = post(url, payload: payload)
```
Why not inline payload?

will-in-wi commented on lib/acme/client.rb:
```diff
@@ -41,93 +39,226 @@ def initialize(jwk: nil, private_key: nil, endpoint: DEFAULT_ENDPOINT, directory
+
+  def certificate(url:)
+    response = download(url)
+    response.body
```
This could be a one-liner `download(url).body`

will-in-wi commented on lib/acme/client.rb:
```diff
@@ -41,93 +39,226 @@ def initialize(jwk: nil, private_key: nil, endpoint: DEFAULT_ENDPOINT, directory
+
+  def deactivate_authorization(url:)
+    payload = { status: 'deactivated' }
+    response = post(url, payload: payload)
```
Maybe inline `payload`?

will-in-wi commented on lib/acme/client.rb:
```diff
@@ -41,93 +39,226 @@ def initialize(jwk: nil, private_key: nil, endpoint: DEFAULT_ENDPOINT, directory
+
+  def request_challenge_validation(url:, key_authorization:)
+    payload = { 'keyAuthorization' => key_authorization }
+    response = post(url, payload: payload)
```
Maybe inline `payload`?

will-in-wi commented on lib/acme/client/faraday_middleware.rb:
```diff
@@ -35,6 +37,12 @@ def on_complete(env)
+  def jws_header
+    headers = { nonce: pop_nonce, url: env.url.to_s }
+    headers[:kid] = client.kid if @mode == :kid
+    headers
```
Style preference, but you could write this as:

```ruby
{ nonce: pop_nonce, url: env.url.to_s }.tap do |headers|
  headers[:kid] = client.kid if @mode == :kid
end
```

unixcharles replied:
I think its a bit more readable to just return the variable

will-in-wi commented on lib/acme/client/jwk/base.rb:
```diff
@@ -56,12 +56,13 @@ def thumbprint
+      alg: jwa_alg
+    }.merge(header)
+    jws[:jwk] = to_h if header[:kid].nil?
+    jws
```
Similarly, you could do:
```ruby
{
  typ: 'JWT',
  alg: jwa_alg
}.merge(header).tap do |jws|
  jws[:jwk] = to_h if header[:kid].nil?
end
```

will-in-wi commented on lib/acme/client/resources/authorization.rb:
```diff
@@ -1,44 +1,86 @@
-
-      challenge.assign_attributes(attributes) if challenge
-    end
+  def initalize_challenge(attributes)
```
`initalize_challenge` should probably be `initialize_challenge`

will-in-wi commented on lib/acme/client/resources/authorization.rb:
```diff
@@ -1,44 +1,86 @@
+  def http01
+    @http01 = challenges.find { |challenge|
+      challenge.is_a?(Acme::Client::Resources::Challenges::HTTP01)
+    }
```
There is a mix of styles using `do; end` vs. `{ }`. Should this and `#dns01` use `do; end`?

unixcharles replied:
There is a bug here because I intended to do `||=`.

On this project I've been following the convention of using `{ }` when the return value and `do / end` when the side effect is desired.

will-in-wi commented on lib/acme/client/resources/challenges/base.rb:
```diff
@@ -1,39 +1,43 @@
-  def client
-    authorization.client
+  def challenge_type
+    self.class::CHALLENGE_TYPE
```
Where is `CHALLENGE_TYPE` set? I don't see `Acme::Client::Resources::Challenges::Base::CHALLENGE_TYPE` defined…

unixcharles replied:
`Base` is an abstract class ,`CHALLENGE_TYPE` is expected to be defined in the child classes. This is also why I do the `self.class::` dance because otherwise it would attempt to find it in the scope of `Base`.

will-in-wi replied:
Ah, that totally makes sense.

will-in-wi commented on spec/spec_helper.rb:
```diff
@@ -1,6 +1,11 @@
 
+require 'openssl'
+
+$directory_url = ENV['ACME_DIRECTORY_URL'] || 'https://127.0.0.1/directory'
```
Could this be a method instead of a global? It totally works but the mere existence of a global feels dirty to me. At least a global method can't be mutated.

will-in-wi commented on spec/support/ssl_helper.rb:
```diff
@@ -26,7 +26,8 @@ def generate_key
         generate_ecdsa_key('secp384r1')
       when 3
-        generate_ecdsa_key('secp521r1')
+        # generate_ecdsa_key('secp521r1')
```
Is this comment needed?

unixcharles replied:
These keys where failing against staging when I started developing. I'll give it another go.

@cpu do you know anything about that?

cpu replied:
@unixcharles Is this something that was working and broke with your V2 work? AFAIK we've never supported secp521r1: https://github.com/letsencrypt/boulder/blob/fa5c9176655d9fa8dfca188de08bd5373aca422f/goodkey/good_key.go#L183-L195

cpu replied:
@unixcharles I don't believe secp521r1 has ever been supported by Let's Encrypt.

unixcharles commented:
Thanks @will-in-wi for the feedback. I've fixed / reply most comment.

weppos commented on README.md:
````diff
@@ -21,135 +21,169 @@ gem 'acme-client'
-# You may need to agree to the terms of service (that's up the to the server to require it or not but boulder does by default)
-registration.agree_terms
+```ruby
+client = Acme::Client.new(private_key: private_key, directory: 'https://acme-v01.api.letsencrypt.org/directory', kid: 'https://example.com/acme/acct/1')
````
Question for the sake of my understanding: I read the ACME v2 specs, and I wonder... is `kid` simply a custom string identifier?

weppos replied:
What about an existing account? Can I fetch it in some ways?

cpu replied:
@weppos From a protocol standpoint there is a way. You can use the mechanism described in [Section 7.3.1](https://tools.ietf.org/html/draft-ietf-acme-acme-10#section-7.3.1) and POST the `newAccount` endpoint sending a signed message body `{"onlyReturnExisting": true}`. The JWS that authenticates the request should embed the JWK of the account key you want to learn the key ID for. If the key is known on the server side you'll get back the KeyID.

From a `unixcharles/acme-client` perspective I don't know, sorry :-) @unixcharles probably does!

unixcharles replied:
> From a unixcharles/acme-client perspective I don't know, sorry :-) @unixcharles probably does!

If the `kid` is not provided at initialization, the client will lazy load the `kid` by doing a `POST` to `newAccount` whenever the `kid` is required.

So you can just instantiate  a new client with the private key without the `kid` and call `Acme::Client#kid` on it to retrieve it.

Good input, I'll add instruction for this.

weppos commented on README.md:
````diff
@@ -21,135 +21,169 @@ gem 'acme-client'
+
+
+```ruby
+client = Acme::Client.new(private_key: private_key, directory: 'https://acme-v01.api.letsencrypt.org/directory')
````
Given this is targeting v2, the directory seems incorrect.

I would actually suggest to use the staging directory in the README as from my experience there is a risk users will just follow the readme blindly using production, without understanding there are strict limits there (and staging should be used for testing).

```
https://acme-staging-v02.api.letsencrypt.org/directory
```

davidcornu commented on README.md:
```diff
@@ -21,135 +21,169 @@ gem 'acme-client'
+  * [Ordering a certificate](#ordering-a-certificate)
+  * [Completing an HTTP challenge](#preparing-for-http-challenge)
+  * [Completing an DNS challenge](#preparing-for-dns-challenge)
+  * [Request a challenge verification](#request-a-challenge-verification)
```
"Requesting" would be more consistent

davidcornu commented on README.md:
```diff
@@ -21,135 +21,169 @@ gem 'acme-client'
-# If authorization.status returns 'valid' here you can already get a certificate
-# and _must not_ try to solve another challenge.
-authorization.status # => 'pending'
+For test purposes you can just save the challenge file and use Ruby to serve it:
```
testing purposes

davidcornu commented on README.md:
```diff
@@ -21,135 +21,169 @@ gem 'acme-client'
 
-# You are not required to send a Content-Type. This method will return the right Content-Type should you decide to include one.
-challenge.content_type
+Once you are ready to complete the challenge, you can request the server perform the verification.
```
you can request that the server perform a verification

davidcornu commented on README.md:
```diff
@@ -21,135 +21,169 @@ gem 'acme-client'
 
-# Optionally save the authorization URI for use at another time (eg: by a background job processor)
-File.write('authorization_uri', authorization.uri)
+You can poll until its status change.
```
changes

davidcornu commented on README.md:
```diff
@@ -21,135 +21,169 @@ gem 'acme-client'
 
-# Wait a bit for the server to make the request, or just blink. It should be fast.
-sleep(1)
+A CSR can be slightly tricky to generate using OpenSSL from Ruby standard library. `acme-client` provide a utility class `CertificateRequest` to help with that.
```
- the Ruby standard library
- provides a utility class (`CertificateRequest`)

davidcornu commented on README.md:
```diff
@@ -21,135 +21,169 @@ gem 'acme-client'
+csr = Acme::Client::CertificateRequest.new(private_key: private_key, subject: { common_name: 'example.com' })
+order.finalize(csr: csr)
+sleep(1) while order.status == 'processing'
+order.certificate # => PEM format certificate
```
PEM-formatted certificate

etiennebarrie commented on README.md:
```diff
@@ -21,135 +21,169 @@ gem 'acme-client'
+## Account management
 
-Before you are able to obtain certificates for your domain, you have to prove that you are in control of it.
+Account are tied to a private key. Before being allowed to create orders, the account must be registered and the ToS accepted using the private key. The account will be assigned a key ID.
```
*Accounts

etiennebarrie commented on README.md:
```diff
@@ -21,135 +21,169 @@ gem 'acme-client'
+The returned order will contain a list of `Authorization` that need to be completed in other to finalize the order, generally one per identifier.
 
-# The http-01 method will require you to respond to a HTTP request.
+Each authorization contains multiple challenges, typically a `dns-01` and a `http-01` challenge. The applicant is only required to complete one of the two challenges.
```
I think we could leave `two` out here.

etiennebarrie commented on lib/acme/client.rb:
```diff
@@ -41,93 +39,221 @@ def initialize(jwk: nil, private_key: nil, endpoint: DEFAULT_ENDPOINT, directory
+    @directory['newOrder'] = URI(body['newOrder']) if body['newOrder']
+    @directory['newAuthz'] = URI(body['newAuthz']) if body['newAuthz']
+    @directory['revokeCert'] = URI(body['revokeCert']) if body['revokeCert']
+    @directory['keyChange'] = URI(body['keyChange']) if body['keyChange']
```
Maybe a loop here?

davidcornu replied:
You could also move this to a `Directory` class that has methods that return URIs for each of these operations. It could also encapsulate the logic in `endpoint_for` and `fetch_directory`

unixcharles replied:
Move it in an object. I was not sure but its pretty nice actually.

etiennebarrie commented on lib/acme/client/error.rb:
```diff
@@ -1,16 +1,32 @@
+
+  class ClientError < Acme::Client::Error; end
+  class InvalidDirectory < ClientError; end
+  class UnsupporedOperation < ClientError; end
```
typo: `UnsupportedOperation`, also used in `endpoint_for`

davidcornu commented on lib/acme/client/resources/account.rb:
```diff
@@ -0,0 +1,58 @@
+    term_of_service = attributes['termsOfServiceAgreed']
+    contact = attributes.fetch('contact', [])
+
+    { term_of_service: term_of_service, status: status, contact: contact }
```
Any reason this doesn't return an instance of `Acme::Client::Resources::Account`?

Seems odd to me to have callers get arguments from a class method to then initialize it separately.

unixcharles replied:
I want each object to be easy serialize/deserialize.

i.e.; I always want to be able to do `ResourceObject.new(client, ResourceObject.to_h)`

Now that you bring it up, I feel like extracting the argument from the response should live in the client instead of the resource objects.

davidcornu commented on lib/acme/client.rb:
```diff
@@ -41,93 +39,221 @@ def initialize(jwk: nil, private_key: nil, endpoint: DEFAULT_ENDPOINT, directory
+    response = post(endpoint_for('newAccount'), payload: payload, mode: :jws)
+    @kid = response.headers.fetch(:location)
+
+    if response.body.nil? || response.body.empty?
```
https://tools.ietf.org/html/draft-ietf-acme-acme-11#section-7.3.1

If an account is created, the server responds with a `201`, if not a `200`.

unixcharles replied:
I know its a bit weird but [pebble](https://github.com/letsencrypt/pebble) and [boulder](https://github.com/letsencrypt/boulder) are no implemented exactly the same. If I remember correctly pebble return empty body on `201` and expect you to follow the location.

I found that the presence of the location was consistent whenever the body was empty, so that's why I did this. Ideally, the body should always be present.

unixcharles replied:
^ @cpu I'm getting this right?

cpu replied:
@unixcharles This is in the context of a normal newAccount request correct, not one with `onlyReturnExisting: true`? I don't think this is right if so.

Pebble returns HTTP status `201` **and** the account object for successful `newAccount` requests that create a new account: https://github.com/letsencrypt/pebble/blob/cae1c3522ee7a4871550f4ff9378677e06a4cdf7/wfe/wfe.go#L739

Boulder does the same thing: HTTP status 201 and the account object in the response body: https://github.com/letsencrypt/boulder/blob/3ecf841a3a48b7c504f26e5b040fbfb36c1664b7/wfe2/wfe.go#L531

unixcharles replied:
I think that it happened when attempting to call `newAccount` on an account that was already registered without `onlyReturnExisting: true`.

I'll try to reproduce, maybe I was just doing something wrong with my test setup.

cpu replied:
> I think that it happened when attempting to call newAccount on an account that was already registered without onlyReturnExisting: true.

That would make sense :+1: - There should not be an HTTP Status Created response in this case because no account was created. That also matches what Section 7.3.1 says when you're not using `onlyReturnExisting`:

     If the server already has an account registered with the provided
     account key, then it MUST return a response with a 200 (OK) status
     code and provide the URL of that account in the Location header
     field.  This allows a client that has an account key but not the
     corresponding account URL to recover the account URL.

In that specific case Boulder returns no body and HTTP status OK: https://github.com/letsencrypt/pebble/blob/cae1c3522ee7a4871550f4ff9378677e06a4cdf7/wfe/wfe.go#L686-L693 Pebble does the same: https://github.com/letsencrypt/pebble/blob/cae1c3522ee7a4871550f4ff9378677e06a4cdf7/wfe/wfe.go#L686-L693

unixcharles replied:
Okay, so it would make sense to follow the location headers if not body is returned.

davidcornu commented on lib/acme/client/error.rb:
```diff
@@ -1,16 +1,32 @@
+  class Caa < ServerError; end
+  class Dns < ServerError; end
+  class Connection < ServerError; end
+  class Tls < ServerError; end
```
nit - Some of these (`Malformed`, `Caa`, `Dns`, `Tls`) are very ambiguous

unixcharles replied:
I map directly to the server error here. I'm not too worried because it come with a clear error message.

davidcornu commented on lib/acme/client/faraday_middleware.rb:
```diff
@@ -67,14 +75,14 @@ def error_name
 
   def error_type_to_klass(type)
-    type.gsub('urn:acme:error:', '').split(/[_-]/).map { |type_part| type_part[0].upcase + type_part[1..-1] }.join
+    type.gsub('urn:ietf:params:acme:error:', '').split(/[_-]/).map { |type_part| type_part[0].upcase + type_part[1..-1] }.join
```
1. I feel like this logic should `live in lib/acme/client/error.rb`
2. A simple `case` on whatever comes after `urn:ietf:params:acme:error:` would be much simpler to reason about (and would let us have less ambiguous error class names https://github.com/unixcharles/acme-client/pull/134/files#r180874585)
3. If you're against (2), 
    - `gsub` &rarr; `sub` would make it clear there's only one substitution going on
    - `sub(/\Aurn:ietf:params:acme:error:/)` is a bit more intention-revealing seen as `urn:ietf:params:acme:error:` should always be at the beginning

unixcharles replied:
Good point. When I came up with this there was no error list in the spec. So I just `#constantize` the error message and kinda defined errors has they appear. I'll replace this with a simple hash in the Error module.

davidcornu commented on lib/acme/client/resources/account.rb:
```diff
@@ -0,0 +1,58 @@
+  end
+
+  def deactivate
+    assign_attributes **@client.account_deactivate.to_h
```
If I'm reading this correctly,

1. `@client.account_deactivate` calls `arguments_from_response`
2. The returned hash is used to initialize a `Acme::Client::Resources::Account`
3. We then call `to_h` on that instance to merge those attributes into this instance

This sounds to me like the responsibilities of the `Account` class are spread all over the place and `@client` is doing too much.

unixcharles replied:
I'm trying to find a way to allow:

```ruby
client.account_deactivate(url: account_url)
```

But also allow:

```ruby
account = client.account(url: account_url)
account.deactivate
```

But without repeating myself too much. I'm not too worried about the `Client` class getting big, as long as it only do http stuff.

I think that exacting the argument from the response `arguments_from_response` is the flaw in this design, that's what make `account > client > account > client > account`. Do that make sense to you?

unixcharles replied:
I've refactor this [here](https://github.com/unixcharles/acme-client/pull/134/files#diff-e843384610a1a0cb8c21582bbed52188R205)

davidcornu commented on lib/acme/client/resources/authorization.rb:
```diff
@@ -1,44 +1,86 @@
+    true
+  end
+
+  def challenges
```
Given the logic in `http01` and `dns01` I'd recommend introducing a `Challenges` object so that we can do

```ruby
def assign_attributes(...)
  @challenges = Challenges.new(challenges)
end
```

which would let you do

```ruby
authorization.challenges.http01
```

unixcharles replied:
Its a nice looking API but I'm not sure if it would bring value. In most scenario you will want to access a specific challenge. i.e.; `authorization.http01`

unixcharles commented:
Thanks for trying it out @nhinds 

I fixed the `Order#reload` bug.

Added `Authorization#identifier` and `Authorization#domain`. This use case just didn't cross my mind since I always do one domain per order/certificate but its pretty legit. The identifier is part of the response but I just didn't assign it anywhere.

@cpu do you happen to know why `"identifier": { "type": "dns", "value": "example.com" }` and not just `"identifier": "example.com"`? Do you know of a case where another type would be possible?

I just assumed its always going to be a `dns` and added a `authorization#domain` attribute to avoid having to dig into the identifier. 

> Also - is the intended way to access the server's terms of service (before agreeing to them) to access client.directory.meta['termsOfService']? Previously the terms of service URI was available on registration.term_of_service_uri.

You are right. Its a bit weird to have to access the directory directly. I added methods for the known meta entries such a `Client#terms_of_service`.

cpu commented:
> @cpu do you happen to know why "identifier": { "type": "dns", "value": "example.com" } and not just "identifier": "example.com"? 

@unixcharles Protocol extensibility. Boulder/Pebble only implement "dns" type identifiers presently but that may change one day.

> Do you know of a case where another type would be possible?

Follow-up protocol extensions (e.g. `draft-ietf-acme-ip-01`) may define additional identifier types. draft-ietf-acme-ip-01 defines [`"type": "ip"` identifiers](https://tools.ietf.org/html/draft-ietf-acme-ip-01#section-3) as one concrete example.

nhinds commented:
Thanks - order reloading works and the `.domain` accessor on authorizations is useful.

`order.certificate` seems to function a bit strangely to me. Previously there was a `Acme::Client::Certificate` class which provided separate access to the certificate and its chain. `order.certificate` currently only returns a string containing the PEM-encoded and its chain together, although it's documented in the README as just a "PEM-formatted certificate".

The certificate is downloaded by sending an `Accept: application/pkix-cert` header. The Acme v2 spec indicates that clients can use this header to request a DER-encoded single certificate, although the current servers ignore this and return a response content type `application/pem-certificate-chain` with the full chain (which is valid, as alternative representations are optional and the full chain is the default return value). If any server does eventually implement the `application/pkix-cert` content type, `order.certificate` will suddenly start returning the DER-encoded rather than the PEM-encoded certificate chain.

If the intention is to return the full chain, could `acme-client` pass `Accept: application/pem-certificate-chain` to indicate that's what it is expecting, so that when servers implement `application/pkix-cert` the method does not break?

It would also be nice to continue to provide separate access to the certificate and its chain, although users of the library can probably migrate to doing so themselves e.g. `certificate, chain = order.certificate.split(/(?<=-----END CERTIFICATE-----)/, 2).map(&:strip)`

unixcharles commented:
> The certificate is downloaded by sending an Accept: application/pkix-cert header. 

Good call, I'll fix the header.

I think that in most scenario, a pem chain is probably what most user need, and the spec make it the default format Its also easy to parse and convert into other format if needed, so I prefer to just return the PEM directly and avoid putting more responsibility on the client.

Does anybody think there would be a good case for wrapping the response?

unixcharles commented:
Thanks everyone for the review and feedback
